### DISCUSSION TITLE
[FEATURE REQUEST] Generate a loop or condition without jinja to allow a more dynamic code ?

### DISCUSSION THREAD

**Discussion by gigi206** [saltstack/salt#67469] [Feature Requests]
**Description**
Jinja implementation does not allow to compute some modules after a specific state because Jinja is computed first at the beginning and that causes some issues.

**Setup**
A simple example, some packages install some users and in this case, Jinja cannot detect the **sshd** user installed by the **openssh-server** package because jinja translate the jinja first
```
openssh-pkg:
  pkg.installed:
    - name: openssh-server
 
// Some code

{% for user in salt['user.getent']() %}
  {% if user['uid'] < 1000 %}
change_user_shell_{{ user['name'] }}:
  user.present:
    - name: {{ user['name'] }}
    - shell: /bin/false
  {% endif %}
{% endfor %}
```

I wonder if the only way is to develop a custom state but develop a custom state each time I encounter this kind of issue is problematic.

**Comment by OrangeDog** [comment]
Just some notes:
If you set every system user's shell to `/bin/false` indiscriminately then you may break something.
If you target specific users rather than looping over all, you can do this with a [requisite](https://docs.saltproject.io/en/latest/ref/states/requisites.html).
You can probably do close to what you want with [slots](https://docs.saltproject.io/en/latest/topics/slots/index.html).

**Comment by gigi206** [comment]
Thank you for your reply :) 

This is just a demonstrative example.

Yes slots are very good but I think it doesn't work in this case because I think it is not possible to loop on a slot.

It missing **with_items** like ansible for that:
```
change_user_shell_{{ item['name'] }}:
  user.present:
    - name: {{ user['name'] }}
    - shell: /bin/false
    - with_items: __slot__:salt:user.getent() 
    - onlyif: {{ item['uid']}} < 1000 
```

**Comment by danielrobbins** [comment]
@gigi206 you are right -- the evaluation of jinja is bound to the .sls file and happens as a pre-processing step, so if you don't use slots, you have to assume that all jinja is evaluated first.

A fairly simple workaround would be to have your first state in its own .sls file, and the second jinja-dependent state in a separate .sls file. Execute the first .sls first, followed by the second .sls. You would essentially be breaking your states into two phases -- one to install the packages, then fix the users. Since the second .sls would have the jinja evaluated at the time it is run, but at this time, any users would be created. This works because of the 'jinja pre-processes by *file*' rule.

When you say that the workaround is to create a 'custom state' -- do you mean a .py state module? If you mean this, then I think this workaround I described is a lot simpler!

Let me know if this is helpful or if you feel this is an acceptable solution, or not.

**Comment by gigi206** [comment]
> You would essentially be breaking your states into two phases -- one to install the packages, then fix the users

I'm not sure to understand, I try to reproduce but with files instead of users (same kind of problem).

I created 4 files:
  - /srv/salt/test/test.jinja
  - /srv/salt/test/test.sls
  - /srv/salt/test/test1.sls
  - /srv/salt/test/test2.sls

**/srv/salt/test/test.jinja**:
```
{% set files = salt['file.find'](path="/tmp/test/*") %}
```

**/srv/salt/test/test.sls**:
```
include:
  - test/test1
  - test/test2
```

**/srv/salt/test/test1.sls**:
```
/tmp/test:
   file.directory:
      - clean: True


/tmp/test/test-{{ salt['random.get_str'](5) }}:
  file.managed:
    - contents: this is a test
```

**/srv/salt/test/test2.sls**:
```
{%- from 'test/test.jinja' import files with context %}

list_files:
  cmd.run:
    - name: echo "{{ files }}"
```

And I run:
```
$ salt-call --local state.apply test/test
local:
----------
          ID: /tmp/test
    Function: file.directory
      Result: True
     Comment: Files cleaned from directory /tmp/test
     Started: 22:57:45.655143
    Duration: 1.377 ms
     Changes:   
              ----------
              /tmp/test/test-ZwnNB:
                  ----------
                  removed:
                      Removed due to clean
              removed:
                  - /tmp/test/test-ZwnNB
----------
          ID: /tmp/test/test-cSz16
    Function: file.managed
      Result: True
     Comment: File /tmp/test/test-cSz16 updated
     Started: 22:57:45.656643
    Duration: 8.451 ms
     Changes:   
              ----------
              diff:
                  New file
----------
          ID: list_files
    Function: cmd.run
        Name: echo "['/tmp/test/test-ZwnNB']"
      Result: True
     Comment: Command "echo "['/tmp/test/test-ZwnNB']"" run
     Started: 22:57:45.666520
    Duration: 6.674 ms
     Changes:   
              ----------
              pid:
                  8999
              retcode:
                  0
              stderr:
              stdout:
                  ['/tmp/test/test-ZwnNB']

Summary for local
------------
Succeeded: 3 (changed=3)
Failed:    0
------------
Total states run:     3
Total run time:  16.502 ms
```

But you can see I see the result of the previous file, not the new one because jinja is evaluated first, indeed stdout show **/tmp/test/test-ZwnNB** instead of the new file **/tmp/test/test-cSz16**.

> When you say that the workaround is to create a 'custom state' -- do you mean a .py state module? If you mean this, then I think this workaround I described is a lot simpler!

Yes that's it

**Comment by danielrobbins** [comment]
@gigi206 The most simple way to address your problem is -- if you need to use jinja, then do 100% jinja. For example, change the first state to use the pkg.install exec function, rather than the state. Or call rpm, etc. directly. That will ensure it is all executed in order, since it is all getting processed in-order as jinja.

And yes, this is a workaround for a missing feature that you describe. If we had a generic means to loop over resources that met certain criteria, we could then apply states to each that match. Typically, if this functionality is needed, it is added to the state itself, so it is more powerful, and you can then list the criteria in a property of the state. This way, there is a state-specific functionality for this.

The problem with state-specific functionality is -- if it doesn't exist, then you are stuck, until a developer adds this functionality. Or you could write a custom state or execution function. Which is of course a lot of work to just get the basic functionality you describe.

The cmd state has some functionality kind of similar to this:
https://docs.saltproject.io/en/latest/ref/states/all/salt.states.cmd.html

I think I agree with you. This is clearly missing functionality that should be added to salt. Some semi-generic bulk operations based on criteria would be very useful.

**Comment by gigi206** [comment]
> if you need to use jinja, then do 100% jinja

This is impossible bacause I manage more than 1000 states called by my top file.  I encounter this issue (this lack of feature) multiple times in my code

> And yes, this is a workaround for a missing feature that you describe. If we had a generic means to loop over resources that met certain criteria, we could then apply states to each that match. Typically, if this functionality is needed, it is added to the state itself, so it is more powerful, and you can then list the criteria in a property of the state. This way, there is a state-specific functionality for this.

Yes it's necessary to loop in a SLS. We could already add conditionnal with onlyif/unless, and now the loop must be added and I hope this feature will be release very soon because developping additionnal states only for do a loop is not very pleasant :(

**Comment by danielrobbins** [comment]
I agree, it should be easier to iterate over a bunch of items and then apply a state to anything matching a condition. I share your perspective @gigi206 -- this should be added as it is a common use case. Iterating over a series of elements and applying a state to anything matching a condition is a very common need. Right now, jinja and states are separated by a wall, as jinja is evaluated in the rendering phase, and states afterwards. Which means that right now, if you want to perform coordinated actions with jinja and states, it is hard. I think the solution is to add some iteration features to states, so you don't need jinja for this.

Here is a suggested implementation for a future version of salt, using a proposed 'runtime template' feature:

```
change_user_shell_mass_edit:
  iterate:
    for: |
      user in salt['user.getent']()
    condition: |
      user['uid'] < 1000
    apply:
      user.present:
         fields:
           raw:
              - shell: /bin/false
           jinja:
               - name: |
                   user['name']
 ```

This is just a suggested possible implementation. The intention here is that this mass_edit rule could use jinja, but it would be evaluated on state runtime rather than as a pre-processor (renderer). The ``for`` condition could evaluate ``salt['user.getent()']``, and for anything where ``user['uid'] < 1000`` is true, it would then use data in the ``apply`` section to dynamically generate a state. The state could have a combination of raw fields as well as fields that come from evaluation of the object in the iterator.

Ultimately, jinja can evaluate it all, but not via being enclosed in ``{{ }}``. Instead, the ``iterate`` state would grab its fields and hand it to jinja to evaluate behind the scenes, and then perform the actions and generate and execute the necessary states.

I suggest using jinja behind-the-scenes so a mass-edit feature still uses the same syntax as our jinja pre-processing commands, so we don't have two different ways of writing things. It should then be easier to move between "pre-processed jinja" and "real-time jinja using the ``iterate`` state." as it is essentially the same code -- if you want it to be evaluated by the pre-processor first, use ``{{ }}``, otherwise use raw jinja without ``{{ }}`` and the ``iterate`` state.

**Comment by techhat** [comment]
I don't know if you're married to using the default `jinja|yaml` renderer, but if you don't mind switching to the `py` renderer maybe this would do it for you:
```
#!py
# This file would be called openssh-pkg.sls, for instance, instead of openssh-pkg.py

def run():
    result = __salt__['state.single'](
        'pkg.installed',
        name='openssh-pkg',
        pkgs=['openssh-server'],
    )

    for state in result:
        if result[state]['result'] is False:
            return {}

    ret = {}
    for user in __salt__['user.getent']():
        if user['uid'] < 1000:
            ret[f"change_user_shell_{user['name']}"] = {
                'user.present': [
                    {'name': user['name']},
                    {'{shell': '/bin/false'},
                ]
            }

    return ret
```
One con here is that you wouldn't see the output from the `pkg.installed` state, but the pro is that the `for` loop wouldn't execute until after the the `pkg.installed` state was run.

**Comment by danielrobbins** [comment]
Thanks @techhat, I think that is a good workaround. I still do think we need a solution like I proposed. The py renderer adds a lot of complexity to what should be a much simpler task, and I think this is such a common use case that we should make it as simple and as intuitive as possible.

**Comment by gigi206** [comment]
I had already try to use the renderer `jinja|py` but it was buggy when I try. There are several open issues with this renderer :(

And I think it can be used as workaround but not for replace the lack of feature

**Comment by danielrobbins** [comment]
@gigi206 take a look at my proposed new feature to make this work intuitively. Would a feature like this solve the problem?

```
change_user_shell_mass_edit:
  iterate:
    for: |
      user in salt['user.getent']()
    condition: |
      user['uid'] < 1000
    apply:
      user.present:
         fields:
           raw:
              - shell: /bin/false
           jinja:
               - name: |
                   user['name']
```

**Comment by gigi206** [comment]
@danielrobbins yes it seems good, I modify a little bit:

```
change_user_shell_mass_edit:
  iterate:
    for: user in salt['user.getent']()
    onlyif:
      - __loop__(user['uid'] < 1000) 
    apply:
      user.present:
          - shell: /bin/false
          - name: __loop__(user['name']) 
```

**Comment by danielrobbins** [comment]
I can probably work on a prototype of this and we can see how it works.
